# <span style='color:brown'> What apps are more likely to attract users ?

![Image](https://bddf794624247cea6a0b-b4761d2ba0154d0278c36dbf2b3c114d.ssl.cf1.rackcdn.com/icons1529553323730.png)

For this project, i will pretend working as a data analyst for a company that builds Android and iOS mobile apps. Let's call that company 'AppStudio'

AppStudio only build apps that meets certain criterias :<br>
* Free to download
* Easy to install

AppStudio makes profit by showing ads to the users inside of its apps. 
So,the more users, the more they can engage with the ads,the more profit AppStudio makes .

My goal for this project is to help the developers understand **what type of apps are more likely to attract users.**

To do so, i will analyse two data sets :
* [Apple store](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) data set with 7000+ apps 
* [Google play](https://www.kaggle.com/lava18/google-play-store-apps)  store data set with 9000+ apps 

## <span style='color:darkblue'> OPENING THE DATA SETS </span>

## Opening of Apple store data set

We open the data set, read it and transform it into a list of list

In [1]:
opened_file_AppleStore = open('AppleStore.csv') # opening
from csv import reader # import the reader module from csv
read_file_AppleStore = reader(opened_file_AppleStore) # read the file
apps_data_AppleStore = list (read_file_AppleStore) # transform it into a list of list

## Opening of Google play data set

We do the same as above

In [2]:
opened_file_googleplaystore = open('googleplaystore.csv') # opening
from csv import reader # import the reader module from csv
read_file_googleplaystore = reader(opened_file_googleplaystore) # read the file
apps_data_googleplaystore = list (read_file_googleplaystore) # transform it into a list of list

## <span style='color:darkblue'> EXPLORATION OF THE DATA SETS  </span>

We can define a function that will help us explore the rows of our data sets

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

## Exploration of  the Apple store data set

In [4]:
print(apps_data_AppleStore[0]) # print the header

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


See below a description of each column:
* `id` : App ID
*`track_name`: App Name
* `size_bytes`: Size (in Bytes)
* `currency`: Currency Type
* `price`: Price amount
* `ratingcounttot`: User Rating counts (for all version)
* `ratingcountver`: User Rating counts (for current version)
* `user_rating` : Average User Rating value (for all version)
* `userratingver`: Average User Rating value (for current version)
* `ver` : Latest version code
* `cont_rating`: Content Rating
* `prime_genre`: Primary Genre
* `sup_devices.num`: Number of supporting devices
* `ipadSc_urls.num`: Number of screenshots showed for display
* `lang.num`: Number of supported languages
* `vpp_lic`: Vpp Device Based Licensing Enabled

In [5]:
explore_data(apps_data_AppleStore[1:], 0, 5, True) # print the first five rows, header not included.

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


For our purpose (finding apps that are more likely to attract users), some columns can be interesting to analyse closer : <br>

`price` (index 4) : indicate the price of an app. 0.0 = free<br>  
`user_rating` (index 7): indicate the average user rating value for all version of an app <br> 

`prime_genre` (index 11): indicate the genre of an app (games, social networking, Music, etc) <br>
    

## Exploration of the Google play store data set

In [6]:
print(apps_data_googleplaystore[0]) # print the header

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


See below a description of each column:
* `App`: Application name
* `Category`: Category the app belongs to
* `Rating`: Overall user rating of the app 
* `Reviews` : Number of user reviews for the app 
* `Size`: Size of the app 
* `Installs` : Number of user downloads/installs for the app 
* `Type` : Paid or Free
* `Price` : Price of the app (as when scraped)
* `Content Rating` : Age group the app is targeted at - Children / Mature 21+ / Adult
* `Genres` : An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to game and music


In [7]:
explore_data(apps_data_googleplaystore[1:], 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


For our analysis, it can be interesting to study these columns : 
* `Category`: Category the app belongs to
* `Genres` : An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to game and music
* `Rating`: Overall user rating of the app
* `Installs` : Number of user downloads/installs for the app 
* `Price` : Price of the app (as when scraped)
Category/Genre, Price, Rating,

## <span style='color:darkblue'> DATA CLEANING </span>

## Are there any missing data?

Lets investigate to see if there is any rows with missing data.<br> One way to detect missing data is to **compare the length of each row with the length of the header**. Every row that is shorter than the header has missing values. <br> To identify these rows, we can use a combination of the for loop and the len() built in function.

Rows with missing data in Apple store data set :

In [8]:
for row in apps_data_AppleStore[1:]: # we exclude the header
    if len(row) != len(apps_data_AppleStore[0]):
        print(row)
        print("\n") # print an empty line, to make our code more readable
        print("Index postion is:", apps_data_AppleStore.index(row)) # Extract the index for each row with missing data

There is no output when we run the for loop, which means that there is no row with missing data in the Apple store data set

Rows with missing data in Google play store data set :

In [9]:
for row in apps_data_googleplaystore[1:]: # we exclude the header
    if len(row) != len(apps_data_googleplaystore[0]):
        print(row)
        print("\n") # print an empty line, to make our code more readable
        print("Index postion is:", apps_data_googleplaystore.index(row)) # Extract the index for each row with missing data

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Index postion is: 10473


In [10]:
del(apps_data_googleplaystore[10473]) # Delete the row with missing data

## Are there any duplicated rows?

I consider duplicated rows to be rows with the same name 

Duplicated rows in Apple Store data set :

In [11]:
unique_values = [] # Will store rows with uniques values
duplicate_values = [] # Will store rows with duplicated values (same name)

for row in apps_data_AppleStore : 
    name = row[0]
    if name in unique_values : 
        duplicate_values.append(name)
    else :
        unique_values.append(name)
print ("the number of duplicated rows in AppleStore Data is : ", len(duplicate_values))
print('\n')
print ("Here are some example of duplicated rows : ", duplicate_values[0:15])

the number of duplicated rows in AppleStore Data is :  0


Here are some example of duplicated rows :  []


There is no duplicated rows in the Apple Store data set,what about google play store?

In [12]:
unique_values = []
duplicate_values = []

for row in apps_data_googleplaystore : 
    name = row[0]
    if name in unique_values : 
        duplicate_values.append(name)
    else :
        unique_values.append(name)
print ("the number of duplicated rows in google store Data is : ", len(duplicate_values))
print('\n')
print("Here are some example of duplicated rows : ",duplicate_values[:15])

the number of duplicated rows in google store Data is :  1181


Here are some example of duplicated rows :  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


So we see that there is a lot of duplicated apps, we need to figure out witch rows to delete and which one to keep in our data sets.
To do so, lets explore one of the duplicated rows.

In [13]:
for row in apps_data_googleplaystore : 
    name = row[0]
    if name == 'Quick PDF Scanner + OCR FREE':
        print (row)

['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


For the google store apps, we can see that the difference is in the column with index = 3 (number of reviews). I think that the row to keep is the one with the highest number of reviews because it is more recent.

### Removing of duplicated rows

In [14]:
reviews_max = {}

for row in apps_data_googleplaystore[1:] :
    name = row[0]
    n_reviews = float(row[3])
    if name not in reviews_max :
        reviews_max[name] = n_reviews
    if name in reviews_max and reviews_max[name] < n_reviews :
        reviews_max[name] = n_reviews

In [15]:
len(reviews_max)

9659

In [16]:
reviews_max['Quick PDF Scanner + OCR FREE']

80805.0

In [17]:
android_clean = [] # we will store our cleaned data
already_added = [] # we will store app names

In [18]:
for row in apps_data_googleplaystore[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name] and (name not in already_added) :
        android_clean.append(row)
        already_added.append(name)

In [19]:
len(android_clean)

9659

In [20]:
apple_clean = apps_data_AppleStore[1:] # we just remove the header

### Removing non-english apps

According to the American Standard Code for Information Interchange, ord (latin characters) range between 0 and 127. We can use that to detect if the name of an app is in English or in another language. Lets say that a name with more than 3 special characters is not english. By doing so we avoir getting rid of apps name 'Instachat 😜' or 

In [21]:
def english_name_detector(word) : 
    count_special_characters = 0
    for character in word :
        if ord(character) > 127 :
            count_special_characters += 1 
            if count_special_characters > 3 :
                return False  
        
    return True

In [22]:
english_name_detector('Instagram')

True

In [23]:
english_name_detector('A爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [24]:
english_name_detector('Instachat 😜') # our fonction doesn't work with emojis...

True

Now lets use our function to remonve non-english apps in the two data sets

In [25]:
# for Apple Store:

In [26]:
apple_english = []
for row in apple_clean : 
    name = row[1] 
    if english_name_detector(name) == True:
        apple_english.append(row)

In [27]:
print(apple_english[50:52])

[['310738695', 'Zillow Real Estate - Homes for Sale & for Rent', '132632576', 'USD', '0.0', '342969', '88478', '4.5', '4.5', '10.4.5', '4+', 'Lifestyle', '37', '5', '1', '1'], ['305343404', 'Tumblr', '151573504', 'USD', '0.0', '334293', '919', '4.0', '4.0', '8.6', '17+', 'Social Networking', '37', '5', '16', '1']]


In [28]:
print(len(apple_clean) - len(apple_english)) # number of non english apps that we removed 

1014


In [29]:
# For Google play store : 

In [30]:
android_english = []
for row in android_clean : 
    name = row[0] 
    if english_name_detector(name) == True:
        android_english.append(row)

In [31]:
print(android_english[50:52])

[['Offroad Oil Tanker Driver Transport Truck 2019', 'AUTO_AND_VEHICLES', '4.3', '542', '33M', '100,000+', 'Free', '0', 'Everyone', 'Auto & Vehicles', 'July 31, 2018', '4.0', '4.1 and up'], ['Tickets SDA 2018 and Exam from the State Traffic Safety Inspectorate with Drom.ru', 'AUTO_AND_VEHICLES', '4.9', '10479', '33M', '100,000+', 'Free', '0', 'Everyone', 'Auto & Vehicles', 'July 18, 2018', '1.7.1', '4.0 and up']]


In [32]:
print(len(android_clean) - len(android_english)) # number of non english apps that we removed 

45


### Removing of non-free apps

In [33]:
# For Apple store :

In [34]:
apple_free = []
for row in apple_english :
    price = float(row[4])
    if price == 0 :
        apple_free.append(row)

In [35]:
apple_free[:2]

[['284882215',
  'Facebook',
  '389879808',
  'USD',
  '0.0',
  '2974676',
  '212',
  '3.5',
  '3.5',
  '95.0',
  '4+',
  'Social Networking',
  '37',
  '1',
  '29',
  '1'],
 ['389801252',
  'Instagram',
  '113954816',
  'USD',
  '0.0',
  '2161558',
  '1289',
  '4.5',
  '4.0',
  '10.23',
  '12+',
  'Photo & Video',
  '37',
  '0',
  '29',
  '1']]

In [36]:
len (apple_english) - len(apple_free)

2961

In [37]:
len(apple_free)

3222

In [38]:
# For Android : 

In [39]:
android_free = []
for row in android_english :
    price = row[7]
    if price == '0' :
        android_free.append(row)

In [40]:
android_free[:2]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up']]

In [41]:
len (android_english) - len(android_free)

750

In [42]:
len(android_free)

8864

In [43]:
# end of the Data cleaning process

## <span style='color:darkblue'> FREQUENCY TABLES </span>

We want to find app that are profitable in both apple and google store. 
We can select many criteria to say if an app will be profitable, for example : Number of user, number or ratings, Hight rating, etc

In [44]:
# apple , genre = index 11 ,
# google, genre = index 9, category = index 1

In [45]:
def freq_table(dataset, index):
    
    dictionnary = {}
    
    for row in dataset : 
        key = row[index]
        if key in dictionnary :
            dictionnary[key] += 1
        else : 
            dictionnary[key] = 1
            
    for key in dictionnary :
        dictionnary[key] = (dictionnary[key]/len(dataset))*100
     
    return dictionnary

In [46]:
freq_table(android_free, 9)

{'Art & Design': 0.5979241877256317,
 'Art & Design;Creativity': 0.06768953068592057,
 'Auto & Vehicles': 0.9250902527075812,
 'Beauty': 0.5979241877256317,
 'Books & Reference': 2.1435018050541514,
 'Business': 4.591606498194946,
 'Comics': 0.6092057761732852,
 'Comics;Creativity': 0.01128158844765343,
 'Communication': 3.2378158844765346,
 'Dating': 1.861462093862816,
 'Education': 5.347472924187725,
 'Education;Creativity': 0.04512635379061372,
 'Education;Education': 0.33844765342960287,
 'Education;Pretend Play': 0.056407942238267145,
 'Education;Brain Games': 0.033844765342960284,
 'Entertainment': 6.069494584837545,
 'Entertainment;Brain Games': 0.078971119133574,
 'Entertainment;Creativity': 0.033844765342960284,
 'Entertainment;Music & Video': 0.16922382671480143,
 'Events': 0.7107400722021661,
 'Finance': 3.7003610108303246,
 'Food & Drink': 1.2409747292418771,
 'Health & Fitness': 3.0798736462093865,
 'House & Home': 0.8235559566787004,
 'Libraries & Demo': 0.936371841155234

In [47]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Now lets generate and analyse some frequency tables 

In [48]:
display_table(apple_free, 11) # sorted frequency table by 'prime_genre'

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


**TOP 10 most common genre** <br> the most common app is Games (58,16%). Then we have entertainment(7,88%).<br> If we take the top 10 genre, we notice that most of the apps are for enternainment (games, photo & video, social networking, music, etc)<br> Games : 58.16263190564867<br>
Entertainment : 7.883302296710118<br>
Photo & Video : 4.9658597144630665<br>
Education : 3.662321539416512<br>
Social Networking : 3.2898820608317814<br>
Shopping : 2.60707635009311<br>
Utilities : 2.5139664804469275<br>
Sports : 2.1415270018621975<br>
Music : 2.0484171322160147<br>
Health & Fitness : 2.0173805090006205<br>
Productivity : 1.7380509000620732<br>

In [49]:
display_table(android_free, 1) # sorted frequency table by 'category'

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [50]:
display_table(android_free, 9) # sorted frequency table by 'genres'

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

# <span style='color:darkblue'> Average number of user ratings </span>

In [51]:
# For Apple Store

In [52]:
freq_table(apple_free, 11)

{'Social Networking': 3.2898820608317814,
 'Photo & Video': 4.9658597144630665,
 'Games': 58.16263190564867,
 'Music': 2.0484171322160147,
 'Reference': 0.5586592178770949,
 'Health & Fitness': 2.0173805090006205,
 'Weather': 0.8690254500310366,
 'Utilities': 2.5139664804469275,
 'Travel': 1.2414649286157666,
 'Shopping': 2.60707635009311,
 'News': 1.3345747982619491,
 'Navigation': 0.186219739292365,
 'Lifestyle': 1.5828677839851024,
 'Entertainment': 7.883302296710118,
 'Food & Drink': 0.8069522036002483,
 'Sports': 2.1415270018621975,
 'Book': 0.4345127250155183,
 'Finance': 1.1173184357541899,
 'Education': 3.662321539416512,
 'Productivity': 1.7380509000620732,
 'Business': 0.5276225946617008,
 'Catalogs': 0.12414649286157665,
 'Medical': 0.186219739292365}

In [53]:
freq_genre_apple = freq_table(apple_free, 11)

In [54]:
for genre in freq_genre_apple:
    total = 0
    len_genre = 0
    
    for app in apple_free :
        genre_app = app[11]
        if genre_app == genre :
            user_ratings = float(app[5])
            total += user_ratings
            len_genre +=1
    avg_ratings = total/len_genre
    print(genre,':', avg_ratings )


Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [55]:
# For Google play

We first have to remove the commas and + in the installs column (for exemple '100,000+') <br> w can use the .replace( ) method :

In [56]:
a = '100,000+'
b = a.replace(',', '')
print(b)
print(b.replace('+', ''))

100000+
100000


In [57]:
freq_category_android = freq_table(android_free, 1)

In [58]:
freq_category_android

{'ART_AND_DESIGN': 0.6430505415162455,
 'AUTO_AND_VEHICLES': 0.9250902527075812,
 'BEAUTY': 0.5979241877256317,
 'BOOKS_AND_REFERENCE': 2.1435018050541514,
 'BUSINESS': 4.591606498194946,
 'COMICS': 0.6204873646209386,
 'COMMUNICATION': 3.2378158844765346,
 'DATING': 1.861462093862816,
 'EDUCATION': 1.1620036101083033,
 'ENTERTAINMENT': 0.9589350180505415,
 'EVENTS': 0.7107400722021661,
 'FINANCE': 3.7003610108303246,
 'FOOD_AND_DRINK': 1.2409747292418771,
 'HEALTH_AND_FITNESS': 3.0798736462093865,
 'HOUSE_AND_HOME': 0.8235559566787004,
 'LIBRARIES_AND_DEMO': 0.9363718411552346,
 'LIFESTYLE': 3.9034296028880866,
 'GAME': 9.724729241877256,
 'FAMILY': 18.907942238267147,
 'MEDICAL': 3.531137184115524,
 'SOCIAL': 2.6624548736462095,
 'SHOPPING': 2.2450361010830324,
 'PHOTOGRAPHY': 2.944494584837545,
 'SPORTS': 3.395758122743682,
 'TRAVEL_AND_LOCAL': 2.33528880866426,
 'TOOLS': 8.461191335740072,
 'PERSONALIZATION': 3.3167870036101084,
 'PRODUCTIVITY': 3.892148014440433,
 'PARENTING': 0.6

In [59]:
for category in freq_category_android :
    total = 0
    len_category = 0
    for app in android_free:
        category_app = app[1]
        if category_app == category :
            installs = app[5]
            installs = installs.replace(',','')
            installs = installs.replace('+','')
            total += float(installs)
            len_category +=1 
    avg_installs = total/len_category
    print(category, ':', avg_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_